cli/upgrade/upgrade.go

```diff
@@ -0,0 +1,183 @@
+package upgrade
+
+import (
+	_ "embed"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/tarantool/tt/cli/connector"
+	"github.com/tarantool/tt/cli/replicaset"
+	"github.com/tarantool/tt/cli/running"
+)
+
+//go:embed lua/upgrade.lua
+var upgradeLuaScript string
+
+type SyncInfo struct {
+	LSN uint32  `mapstructure:"lsn"`
+	IID uint32  `mapstructure:"iid"`
+	Err *string `mapstructure:"err"`
+}
+
+func internalUpgrade(conn connector.Connector) (SyncInfo, error) {
+	var upgradeInfo SyncInfo
+	res, err := conn.Eval(upgradeLuaScript, []any{}, connector.RequestOpts{})
+	if err != nil {
+		return upgradeInfo, err
+	}
+	if err := mapstructure.Decode(res[0], &upgradeInfo); err != nil {
+		return upgradeInfo, err
+	}
+	if upgradeInfo.Err != nil {
+		return upgradeInfo, errors.New(*upgradeInfo.Err)
+	}
+	return upgradeInfo, nil
+}
+
+func WaitLSN(conn connector.Connector, masterUpgradeInfo SyncInfo, timeout int) error {
+	for i := 0; i < timeout; i++ {
```

Comment:
Wow, this is a non-typical retrying pattern, at least it is surprising for me:

```go
for i := 0; i < timeout; i++ {
    // <..actions..>
    time.Sleep(time.Second)
}
```

I usually use a `double` timeout and something like the following (pseudocode):

```go
var deadline = time.Now() + timeout
for {
    // <..actions..>
    if time.Now() > deadline {
        break // or return error
    }
    time.Sleep(DELAY)
}
```

It performs at least one action disregarding the timeout (even if it is zero) and it accounts a time needs to perform the actions (`Eval` in our case).

Reply:
Replaced with the suggested format.